### PR TITLE
Feature/notes search

### PR DESCRIPTION
**Pull request by vince-a-m**
Added:
New route for /notes (and corresponding dropdown option within the "More" section of the toolbar)
New component "notes-search-list" to be displayed on the /notes route
New clickBehavior type to stixList "linkToObjectRef" (clicking row will link to the first object in the object_refs array and open the notes in the sidebar service)
New stixList row for "linkToObjectRef" clickBehavior (matIcon/tooltip)
New stixList formatting and behavior for Notes object (row display)

Updated:
Look and feel of "dialog" clickBehavior on stixList (new matIcon/tooltip)

### CODE REVIEW

File: app/src/app/components/stix/stix-list/stix-list.component.html [hunk 1]
```diff
@@ -90,15 +90,23 @@
...
-                    <th mat-header-cell *matHeaderCellDef>
-                        <!-- nothing in header -->
-                    </th>
+                <!-- open new dialog icon column (for relationships) -->
+                <ng-container matColumnDef="open-dialog" stickyEnd>
+                    <th mat-header-cell *matHeaderCellDef></th> <!-- nothing in header -->
                     <td mat-cell *matCellDef="let element" class="icon-cell">
-                        <mat-icon aria-hidden="false" aria-label="open">open_in_new</mat-icon>
+                        <mat-icon aria-hidden="false" aria-label="open-dialog"
+                          matTooltip="Info about the action">expand_content</mat-icon>
```
Comment (clemiller) [inline app/src/app/components/stix/stix-list/stix-list.component.html:98]: We may want to keep this as a `open_in_new` icon to keep it consistent with how we've been using it as an `open in dialog` icon.

Also recommend updating the tooltip from

> Info about the action

to

>  View object in dialog

image

File: app/src/app/views/stix/notes-search/notes-search-list/notes-search-list.component.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  selector: 'app-notes-search-list',
+  templateUrl: './notes-search-list.component.html',
+  styleUrls: ['./notes-search-list.component.scss']
+})
+export class NotesSearchListComponent implements OnInit {
+  constructor(private sidebarService: SidebarService) { 
+  }
+
+  ngOnInit(): void {
+    this.sidebarService.opened = false;
```
Comment (clemiller) [inline app/src/app/views/stix/notes-search/notes-search-list/notes-search-list.component.ts:14]: Closing the sidebar here causes an `ExpressionChangedAfterItHasBeenChecked` error to be thrown when the `notes-search` page loads from a previous page where the resource drawer is already opened. The resource drawer should remain open if it's been opened by the user.

image

File: app/src/app/components/stix/stix-list/stix-list.component.ts [hunk 3]
```diff
@@ -437,9 +450,18 @@ export class StixListComponent implements OnInit, AfterViewInit, OnDestroy {
...
         }
+        else if (this.config.clickBehavior && this.config.clickBehavior == "linkToObjectRef") {
+          // technically a note can be linked to many objects, we will select the first object
+          let object_ref = element['object_refs'][0];
+          // Get type to navigate from target_ref
+          let type = this.typeMap[object_ref.split('--')[0]];
+
+          this.sidebarService.opened = true;
+          this.sidebarService.currentTab = 'notes';
+          this.router.navigateByUrl('/'+ type + '/' + object_ref);
```
Comment (clemiller) [inline app/src/app/components/stix/stix-list/stix-list.component.ts:463]: This is a great addition! I like that it opens the sidebar directly to the `notes` tab on the object page. This makes it super easy to view the context.

File: app/src/app/components/stix/stix-list/stix-list.component.ts [hunk 4]
```diff
@@ -437,9 +450,18 @@ export class StixListComponent implements OnInit, AfterViewInit, OnDestroy {
             let target_ref = element['target_ref'];
             // Get type to navigate from target_ref
             let type = this.typeMap[target_ref.split('--')[0]];
-
             this.router.navigateByUrl('/'+ type + '/' + target_ref);
         }
+        else if (this.config.clickBehavior && this.config.clickBehavior == "linkToObjectRef") {
+          // technically a note can be linked to many objects, we will select the first object
+          let object_ref = element['object_refs'][0];
```
Comment (clemiller) [inline app/src/app/components/stix/stix-list/stix-list.component.ts:457]: Good catch! A note can _technically_ be associated with many objects (according to the STIX specification), but our use of it is only one-to-one, so there will only be a single object in the list.

**Review by clemiller — CHANGES_REQUESTED** [review]
This looks good so far - there are a couple of improvements needed:

- [x]  I'd recommend moving the `notes-search` component up a level to the `views/` directory (along side `reference-manager`), rather than the `views/stix/` directory, since it doesn't extend `StixPageComponent`. While it is technically a STIX object, it doesn't require a [StixViewConfig](https://github.com/center-for-threat-informed-defense/attack-workbench-frontend/blob/develop/app/src/app/views/stix/stix-view-page.ts#L23).

- [x] The "notes" option should be moved to underneath the "reference manager" option in the "more" menu.

- [x] In the notes list, clicking on the `open_in_new` icon on the right does not redirect the user to the related object page; the user is only redirected if they click on a different part of the row. If possible, we should make the icon clickable as well, or remove the `cursor:pointer` styling.

- [ ] The latest changes currently staged on `develop` should be pulled into your feature branch to keep everything in sync.

- [x]  An entry for this feature should be added to `CHANGELOG.md` under the `Changes Staged on Develop` for v1.3.1 (this is currently in the `develop` branch). This is just a short description of the changes and a link to #175.

File: app/src/app/components/stix/stix-list/stix-list.component.html [hunk 5]
```diff
@@ -90,15 +90,23 @@
...
                     </td>
                 </ng-container>
+
+                <!-- open link icon column (for notes) -->
+                <ng-container matColumnDef="open-link" stickyEnd>
+                  <th mat-header-cell *matHeaderCellDef></th> <!-- nothing in header -->
+                  <td mat-cell *matCellDef="let element" class="icon-cell">
+                      <mat-icon aria-hidden="false" aria-label="open-link"
+                        matTooltip="Opens the object linked to this object">open_in_new</mat-icon>
```
[the remark this answers is not in the document]
Comment (vince-a-m) [reply app/src/app/components/stix/stix-list/stix-list.component.html]: That's probably the best icon for this.

File: app/src/app/components/stix/stix-list/stix-list.component.ts (hunk 3 above)
Comment (vince-a-m) [reply app/src/app/components/stix/stix-list/stix-list.component.ts:463]: Thanks :)

**Review by clemiller — APPROVED** [review]

**Pull request merged**